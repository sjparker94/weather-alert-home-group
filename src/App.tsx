import React from 'react';
import '@csstools/normalize.css';
import { BrowserRouter as Router, Route } from 'react-router-dom';

import Pages from './pages/Pages';
import PageLayout from './components/PageLayout/PageLayout';

const App: React.FC = () => {
    // const isInitialLoad = useSelector<AppState, boolean>(state => state.locations.isInitialLoad);
    // const app = !isInitialLoad ? <Route component={Pages} /> : null;
    return (
        <Router>
            <PageLayout>
                <Route component={Pages} />
            </PageLayout>
        </Router>
    );
};

export default App;
