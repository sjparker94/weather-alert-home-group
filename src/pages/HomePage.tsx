import React from 'react';
import HomeContentWrapper from '../components/ContentBlock/ContentWrapper';
import SearchForm from '../components/SearchForm/SearchForm';
import ContentBlock from '../components/ContentBlock/ContentBlock';

const HomePage: React.FC = () => {
    return (
        <HomeContentWrapper>
            <h1 className="welcome-heading">Welcome!</h1>
            <p className="welcome-paragraph">
                Use the form below to search for a city and add it to your favourites
            </p>
            <ContentBlock>
                <SearchForm />
            </ContentBlock>
        </HomeContentWrapper>
    );
};

export default HomePage;
