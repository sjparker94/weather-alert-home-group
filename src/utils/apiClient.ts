import axios from 'axios';

const apiClient = axios.create({
    baseURL: process.env.REACT_APP_OPENWEATHER_API_URL,
    responseType: 'json',
    headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
    },
});

export default apiClient;
