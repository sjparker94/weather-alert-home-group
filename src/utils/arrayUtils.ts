export function findByProp<T, K extends keyof T>(prop: K, propToFind: T[K]) {
    return (item: T) => item[prop] === propToFind;
}
