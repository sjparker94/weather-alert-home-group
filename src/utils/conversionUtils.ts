/** metres per second to miles per hour */
export const mpsToMph = (value: number) => Math.round(((value * 3600) / 1610.3) * 1000) / 1000;

/** mph to km/h */
export const mpsToKmh = (value: number) => value * 3.6;

/** celsius to fahrenheit */
export const convertToFahrenheit = (celsius: number) => (celsius * 9) / 5 + 32;
