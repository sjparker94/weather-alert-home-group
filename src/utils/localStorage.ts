import AppState from '../interfaces/AppState';
import Location from '../interfaces/Location';
import { locationsInitialState } from '../reducers/locationsReducer';
import { settingsInitialState } from '../reducers/settingsReducer';

export const loadState = (): AppState | undefined => {
    try {
        const serializedLocations = localStorage.getItem('locations');

        const isFahrenheit = localStorage.getItem('isFahrenheit') === 'true';
        const isKm = localStorage.getItem('isKm') === 'true';

        const locations: Location[] =
            serializedLocations === null
                ? locationsInitialState.data
                : JSON.parse(serializedLocations);

        return {
            locations: {
                ...locationsInitialState,
                data: locations,
            },
            settings: {
                ...settingsInitialState,
                isFahrenheit,
                isKm,
            },
        };
    } catch (err) {
        // undefined will initialise the store with default values
        return undefined;
    }
};

export const saveState = (state: AppState) => {
    try {
        const {
            locations: { data: locations },
            settings: { isFahrenheit, isKm },
        } = state;
        const serializedLocations = JSON.stringify(locations);
        localStorage.setItem('locations', serializedLocations);
        localStorage.setItem('isFahrenheit', JSON.stringify(isFahrenheit));
        localStorage.setItem('isKm', JSON.stringify(isKm));
    } catch {
        // ignore errors
    }
};
