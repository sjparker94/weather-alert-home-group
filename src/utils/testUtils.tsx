import Location from '../interfaces/Location';
import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render as rtlRender } from '@testing-library/react';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';
import rootReducer from '../reducers/index';
import thunk from 'redux-thunk';
import theme from '../styles/theme';

export const fakeLocation = (): Location => ({
    coord: {
        lon: -0.13,
        lat: 51.51,
    },
    weather: [
        {
            id: 802,
            main: 'Clouds',
            description: 'scattered clouds',
            icon: '03d',
        },
    ],
    base: 'stations',
    main: {
        temp: 22.85,
        feels_like: 19.87,
        temp_min: 22.22,
        temp_max: 24.44,
        pressure: 1028,
        humidity: 31,
    },
    visibility: 10000,
    wind: {
        speed: 2.6,
        deg: 0,
    },
    clouds: {
        all: 40,
    },
    dt: 1594560101,
    sys: {
        type: 1,
        id: 1414,
        country: 'GB',
        sunrise: 1594526265,
        sunset: 1594584852,
    },
    timezone: 3600,
    id: 2643743,
    name: 'London',
    cod: 200,
});

const render = (ui: any, initialStore = {}, options = {}) => {
    const store = createStore(rootReducer, initialStore, applyMiddleware(thunk));

    const Providers = ({ children }: any) => (
        <Provider store={store}>
            <ThemeProvider theme={theme}>{children}</ThemeProvider>
        </Provider>
    );

    return rtlRender(ui, { wrapper: Providers, ...options });
};
// re-export everything
export * from '@testing-library/react';

// override render method
export { render };
