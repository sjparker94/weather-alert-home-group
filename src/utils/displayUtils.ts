import { convertToFahrenheit, mpsToKmh, mpsToMph } from './conversionUtils';
import WindSpeedTextValue from '../interfaces/WindSpeedTextValue';

/** calculate which temp value to show based on the settings state
 * @param temp temperature value in celsius
 */
export const getTempDisplayValue = (temp: number, isFahrenheit: boolean) =>
    Math.round(isFahrenheit ? convertToFahrenheit(temp) : temp);

/** calculate which temp value to show based on the settings state
 * @param speed speed value in mps
 */
export const getSpeedDisplayValue = (speed: number, isKm: boolean) =>
    Math.round(isKm ? mpsToKmh(speed) : mpsToMph(speed));

/** Calculate the string value of the wind speed
 * @param speed - speed in mph
 */
export const calculateWindSpeedTextValue = (speed: number): WindSpeedTextValue => {
    if (speed < 5) {
        return 'light';
    }
    if (speed < 12) {
        return 'gentle';
    }
    if (speed < 25) {
        return 'moderate';
    }
    if (speed < 50) {
        return 'strong';
    }
    return 'storm';
};

export function toTextualWindDirection(degree: number) {
    const sectors = [
        'Northerly',
        'North Easterly',
        'Easterly',
        'South Easterly',
        'Southerly',
        'South Westerly',
        'Westerly',
        'North Westerly',
    ];

    degree += 22.5;

    if (degree < 0) {
        degree = 360 - (Math.abs(degree) % 360);
    } else {
        degree = degree % 360;
    }

    const which = Math.floor(degree / 45);
    return sectors[which];
}
