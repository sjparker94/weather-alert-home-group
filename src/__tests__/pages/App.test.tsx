import React from 'react';
import axios from 'axios';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';

// @ts-ignore
import { render, wait, fakeLocation } from '../../utils/testUtils';
import HomePage from '../../pages/HomePage';
import App from '../../App';

jest.mock('axios');

const mockLocation = fakeLocation();

describe('<App />', () => {
    it('renders and matches snapshot', async () => {
        const { container, getByTestId } = render(<App />);

        // Check that the search button is in the dom
        expect(getByTestId('location-search-input')).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    it('handles user input', async () => {
        const { getByPlaceholderText, getByTestId } = render(<App />);

        const inputEl = getByTestId('location-search-input');

        expect(inputEl).toBeInTheDocument();
        await userEvent.type(inputEl, 'London');

        expect(inputEl).toHaveAttribute('value', 'London');
    });

    it('handles user input', async () => {
        const { getByText, getByPlaceholderText } = render(<App />);

        const inputEl = getByPlaceholderText(/Newcastle Upon Tyne/i);

        expect(getByText('Search')).toBeInTheDocument();
        expect(inputEl).toBeInTheDocument();
        await userEvent.type(inputEl, 'London');

        expect(inputEl).toHaveAttribute('value', 'London');
    });

    it('displays confirm block after a successful search', async () => {
        const { getByPlaceholderText, getByTestId, getByText, queryByTestId } = render(<App />);

        const inputEl = getByPlaceholderText(/Newcastle Upon Tyne/i);
        const buttonEl = getByTestId('main-search-button');

        axios.get = jest.fn().mockResolvedValue({ data: mockLocation });

        expect(inputEl).toBeInTheDocument();
        await userEvent.type(inputEl, 'London');

        await userEvent.click(buttonEl);
        await wait(() => {
            // Check the api has been called
            expect(axios.get).toHaveBeenCalledTimes(1);
            // Check any errors are not displaying
            expect(queryByTestId('search-error')).not.toBeInTheDocument();
            // Check that the return from the api is being displayed
            expect(getByText(mockLocation.name)).toBeInTheDocument();
        });
    });

    it('displays error block after a search failure', async () => {
        const { getByPlaceholderText, getByTestId } = render(<App />);

        const inputEl = getByPlaceholderText(/Newcastle Upon Tyne/i);
        const buttonEl = getByTestId('main-search-button');

        axios.get = jest.fn().mockRejectedValue(new Error());

        expect(inputEl).toBeInTheDocument();
        await userEvent.type(inputEl, 'lssdk');

        userEvent.click(buttonEl);

        await wait(() => {
            // Check that the error is being displayed
            expect(getByTestId('search-error')).toBeInTheDocument();
        });
    });

    it('handles the cancelling of a location search', async () => {
        const { getByPlaceholderText, getByTestId, queryByText, queryByTestId } = render(
            <HomePage />
        );

        const inputEl = getByPlaceholderText(/Newcastle Upon Tyne/i);
        const buttonEl = getByTestId('main-search-button');

        axios.get = jest.fn().mockResolvedValue({ data: mockLocation });

        expect(inputEl).toBeInTheDocument();
        await userEvent.type(inputEl, 'London');

        userEvent.click(buttonEl);
        await wait(() => {
            userEvent.click(getByTestId('cancel-search-button'));

            // Check that the confirm block has been removed from the dom
            expect(queryByText(mockLocation.name)).not.toBeInTheDocument();
        });
    });

    it('handles the confirming of a location search and adds to the favourites', async () => {
        const {
            getByPlaceholderText,
            getByTestId,
            getAllByTestId,
            queryByText,
            queryByTestId,
        } = render(<App />);

        const inputEl = getByPlaceholderText(/Newcastle Upon Tyne/i);
        const buttonEl = getByTestId('main-search-button');

        axios.get = jest.fn().mockResolvedValue({ data: mockLocation });

        expect(inputEl).toBeInTheDocument();
        await userEvent.type(inputEl, 'London');

        userEvent.click(buttonEl);
        await wait(() => {
            userEvent.click(getByTestId('confirm-search-button'));

            const favouriteLocationEls: HTMLElement[] = getAllByTestId('favourite-location');

            // Check that the favourite locations have been updated and now contain an item
            expect(favouriteLocationEls.length).toBe(1);
            // Check that the correct location has been added
            expect(favouriteLocationEls[0]).toHaveTextContent(mockLocation.name);
        });
    });
});
