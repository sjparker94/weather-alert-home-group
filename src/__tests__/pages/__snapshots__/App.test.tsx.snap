// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders and matches snapshot 1`] = `
<div>
  <div
    class="sc-fznyAO dYQzzH"
  >
    <div
      class="main-page-layout-wrapper"
    >
      <main
        class="main-page-content"
      >
        <div
          class="sc-AxjAm oEzzX"
        >
          <h1
            class="welcome-heading"
          >
            Welcome!
          </h1>
          <p
            class="welcome-paragraph"
          >
            Use the form below to search for a city and add it to your favourites
          </p>
          <div
            class="sc-AxheI fKbPLh"
            title="Location Search"
          >
            <div
              class="sc-AxgMl kizETN"
            >
              <h3>
                Location Search
              </h3>
            </div>
            <form
              class="sc-AxhUy hPtbI"
            >
              <div
                class="inputs-wrapper"
              >
                <div
                  class="sc-AxiKw icyWgc"
                >
                  <label
                    for="city"
                  >
                    Enter the name of a city
                  </label>
                  <input
                    aria-label="Enter the name of a city"
                    data-testid="location-search-input"
                    id="city"
                    name="city"
                    placeholder="Newcastle upon Tyne"
                    type="text"
                    value=""
                  />
                  <p
                    class="input-description"
                  >
                    E.g. Newcastle Upon Tyne or New York
                  </p>
                </div>
              </div>
              <button
                class="sc-AxirZ gKFdqJ"
                data-testid="main-search-button"
                disabled=""
              >
                <span
                  class="material-icons"
                >
                  search
                </span>
                Search
              </button>
            </form>
          </div>
        </div>
      </main>
      <div
        class="sc-fzozJi jfdNKJ"
      >
        <h2
          class="main-title"
        >
          <span
            class="material-icons star-icon"
          >
            star
          </span>
           Your Favourites
        </h2>
        <div
          class="no-locations-saved"
          data-testid="favourites-no-locations"
        >
          <p>
            No locations have been selected use the form on the home page to add one
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
